core/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExprUtils.scala

```diff
@@ -0,0 +1,221 @@
+package org.apache.spark.sql.catalyst.expressions
+
+import com.pingcap.tikv.expression.visitor.{ColumnMatcher, MetaResolver, SupportedExpressionValidator}
+import com.pingcap.tikv.expression.{AggregateFunction, ByItem, ColumnRef, ExpressionBlacklist}
+import com.pingcap.tikv.meta.{TiColumnInfo, TiDAGRequest, TiTableInfo}
+import com.pingcap.tikv.region.RegionStoreClient.RequestTypes
+import com.pingcap.tispark.TiDBRelation
+import org.apache.spark.sql.catalyst.expressions.aggregate.{AggregateExpression, Average, Count, DeclarativeAggregate, First, Max, Min, PromotedSum, Sum, SumNotNullable}
+import org.apache.spark.sql.execution.TiConverter.fromSparkType
+
+import scala.collection.JavaConversions._
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+object ExprUtils {
+  def transformGroupingToTiGrouping(expr: Expression,
+                                    meta: TiTableInfo,
+                                    dagRequest: TiDAGRequest): Unit =
+    expr match {
+      case BasicExpression(keyExpr) =>
+        MetaResolver.resolve(keyExpr, meta)
+        dagRequest.addGroupByItem(ByItem.create(keyExpr, false))
+        // We need to add a `First` function in DAGRequest along with group by
+        dagRequest.getFields.asScala
+          .filter(ColumnMatcher.`match`(_, keyExpr))
+          .foreach(
+            (ref: ColumnRef) =>
+              dagRequest
+                .addAggregate(
+                  AggregateFunction
+                    .newCall(
+                      AggregateFunction.FunctionType.First,
+                      ref,
+                      meta.getColumn(ref.getName).getType
+                    )
+              )
+          )
+      case _ =>
+    }
+
+  def transformAggExprToTiAgg(expr: Expression, meta: TiTableInfo, dagRequest: TiDAGRequest): Any =
```

Comment:
have you tried aggregation on unsigned values?

Reply:
Yes. Any overflowed values will be prohibit to push down to coprocessor.